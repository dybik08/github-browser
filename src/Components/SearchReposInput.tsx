import React from 'react';
import { Button, Input, message, Tooltip } from 'antd';
import { fetchRepos } from '../actions/networkActions';
import { Repository } from '../Constants/types';
import { InfoCircleOutlined } from '@ant-design/icons/lib';
const { Search } = Input;

const formatReposResponseData = (data: { items: Repository[] }): Repository[] => {
    return data.items.map((repoData: Repository, index: number) => ({
        key: `${index}`,
        login: repoData.owner.login,
        language: repoData.language,
        stargazers_count: repoData.stargazers_count,
        name: repoData.name,
        owner: repoData.owner,
        created_at: repoData.created_at,
        updated_at: repoData.updated_at,
        license: repoData.license,
        description: repoData.description,
        forks: repoData.forks,
    }));
};

interface SearchReposInputProps {
    reposLoading: boolean;
    setReposLoading: (loading: boolean) => void;
    setRepos: (repos: Repository[]) => void;
}

const SearchReposInput: React.FC<SearchReposInputProps> = props => {
    const { setReposLoading, setRepos, reposLoading } = props;
    const [inputValue, setInputValue] = React.useState<string>('');

    const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        setInputValue(event.target.value);
    };

    const onSearchReposButtonPress = async () => {
        if (inputValue.length === 0) {
            return message.error('Please fill search bar before continue', 1);
        }

        const multipleSearch = inputValue.replace('/,/g', '+');

        setReposLoading(true);
        const { data } = await fetchRepos(multipleSearch);
        const formattedData = formatReposResponseData(data);

        setRepos(formattedData);
        setReposLoading(false);
    };

    return (
        <Search
            autoFocus={true}
            className='search-row'
            value={inputValue}
            onSearch={onSearchReposButtonPress}
            onChange={event => handleInputChange(event)}
            placeholder='Search github for repositories'
            loading={reposLoading}
            enterButton
            suffix={
                <Tooltip title='Split text by comma to search for multiple items e.g. redux, react '>
                    <InfoCircleOutlined style={{ color: 'rgba(0,0,0,.45)' }} />
                </Tooltip>
            }
        />
    );
};

export default SearchReposInput;
