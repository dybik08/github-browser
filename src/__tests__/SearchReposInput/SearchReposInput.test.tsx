import React, { FC, ReactElement } from 'react';
import { configure, shallow, ShallowWrapper } from 'enzyme';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import SearchReposInput from '../../Components/SearchReposInput';

configure({ adapter: new Adapter() });

describe('RepoListItem', () => {
    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, 'useState');
    useStateSpy.mockImplementation((init: any) => [init, setState]);
    let wrapper: ShallowWrapper = shallow(
        <SearchReposInput reposLoading={false} setRepos={jest.fn()} setReposLoading={jest.fn()} />
    );

    it('renders correctly', () => {
        const tree = renderer.create(wrapper).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should update state on input change', () => {
        const newInputValue = 'React is Awesome';
        wrapper.find('.search-row').simulate('change', { target: { value: newInputValue } });
        expect(setState).toHaveBeenCalledWith(newInputValue);
    });
});
