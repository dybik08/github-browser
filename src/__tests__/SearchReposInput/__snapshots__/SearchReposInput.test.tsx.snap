// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepoListItem renders correctly 1`] = `
<span
  className="ant-input-group-wrapper ant-input-search ant-input-search-enter-button search-row"
>
  <span
    className="ant-input-wrapper ant-input-group"
  >
    <span
      className="ant-input-affix-wrapper ant-input-search ant-input-search-enter-button search-row"
      onMouseUp={[Function]}
      style={null}
    >
      <input
        autoFocus={true}
        className="ant-input"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        placeholder="Search github for repositories"
        style={null}
        type="text"
        value=""
      />
      <span
        className="ant-input-suffix"
      >
        <span
          aria-label="info-circle"
          className="anticon anticon-info-circle"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          role="img"
          style={
            Object {
              "color": "rgba(0,0,0,.45)",
            }
          }
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="info-circle"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
            />
            <path
              d="M464 336a48 48 0 1096 0 48 48 0 10-96 0zm72 112h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V456c0-4.4-3.6-8-8-8z"
            />
          </svg>
        </span>
      </span>
    </span>
    <span
      className="ant-input-group-addon"
    >
      <button
        className="ant-btn ant-btn-primary ant-input-search-button"
        onClick={[Function]}
        onMouseDown={[Function]}
        type="button"
      >
        <span
          aria-label="search"
          className="anticon anticon-search"
          role="img"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="search"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
            />
          </svg>
        </span>
      </button>
    </span>
  </span>
</span>
`;
